src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/TestDiscoverer.cs

```diff
@@ -0,0 +1,152 @@
+
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Composition;
+using Microsoft.CodeAnalysis.Features.Testing;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Testing;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.Extensions.Logging;
+using Microsoft.TestPlatform.VsTestConsole.TranslationLayer;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Testing;
+
+[Export, Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal class TestDiscoverer(ILoggerFactory loggerFactory)
+{
+    /// <summary>
+    /// TODO - localize messages. https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1799066/
+    /// </summary>
+    private const string StageName = "Discovering tests...";
+    private readonly ILogger _logger = loggerFactory.CreateLogger<TestDiscoverer>();
+
+    /// <summary>
+    /// Finds tests in the specified document in the specified range.
+    ///
+    /// Note that since tests run against the last built dll,
+    /// </summary>
+    public async Task<ImmutableArray<TestCase>> DiscoverTestsAsync(
+        LSP.Range range,
+        Document document,
+        string projectOutputPath,
+        BufferedProgress<RunTestsPartialResult> progress,
+        VsTestConsoleWrapper vsTestConsoleWrapper,
+        CancellationToken cancellationToken)
+    {
+        var partialResult = new RunTestsPartialResult(StageName, $"{Environment.NewLine}Starting test discovery", Progress: null);
+        progress.Report(partialResult);
+
+        var testMethodFinder = document.GetRequiredLanguageService<ITestMethodFinder>();
+
+        // Find any potential test methods (based on attributes) that exist in the input range.
+        var potentialTestMethods = await GetPotentialTestMethodsAsync(range, document, testMethodFinder, cancellationToken);
+        if (potentialTestMethods.IsEmpty)
+        {
+            progress.Report(partialResult with { Message = "No test methods found in requested range" });
+            return ImmutableArray<TestCase>.Empty;
+        }
+
+        // Next, run the actual vs test discovery on the output dll to figure out what tests actually exist.
+        var discoveryHandler = new DiscoveryHandler(progress);
+        var stopwatch = SharedStopwatch.StartNew();
+
+        // The async APIs for vs test are broken (current impl ends up just hanging), so we must use the sync API instead.
+        // TODO - run settings.  https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1799066/
```

Comment:
the linked issue is resolved.  can we make this async again?

Reply:
The issue is actually tracking providing run settings (which I have planned for a followup, see PR description).  The async APIs for vs test are not fixed, we have to use the sync ones.